src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java

```diff
@@ -0,0 +1,361 @@
+                    }
+                    shipping.start();
+                    if (!ret.isEmpty()) {
+                        kving.start();
+                        accepted += ret.size();
+                        KV<String, Iterable<GATKRead>> out = KV.of(currentShard.toString(), ret);
+                        kving.stop();
+                        outputting.start();
+                        c.output(out);
+                        outputting.stop();
+                        //bunny.stepEnd("output " + ret.size() + " reads");
+                    }
+                    shipping.stop();
+                    readsEmitted += accepted;
+                    //bunny.stepEnd("loading " + accepted + " reads for " + i.toString()
+                    //    + " in " + processing.elapsed(TimeUnit.MILLISECONDS) + " ms. "
+                    //    + " readingInput: " + readingInput.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " init: " + init.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " cutShards: " + cutShards.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " kv.of: " + kving.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " output: " + outputting.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " rejected " + rejected);
+                }
+
+                private boolean accept(GATKRead r, SimpleInterval region) {
+                    if (r.isUnmapped()) {
+                        // we don't want unmapped reads, *unless* their mate is in the region
+                        return (!r.mateIsUnmapped() && r.getMateStart() >= region.getStart());
+                    } else {
+                        // the query gives us reads that overlap the region. We want only reads that *start* in those intervals.
+                        return (r.getStart() >= region.getStart());
+                    }
+                }
+
+                private boolean accept(SAMRecord r, SimpleInterval region) {
+                    if (r.getReadUnmappedFlag() ||
+                        r.getReferenceName() == null || r.getReferenceName().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_NAME) ||
+                        r.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START) {
+                        // unmapped read. We don't want it, *unless* their mate is in the region
+                        return (!r.getMateUnmappedFlag() && r.getMateAlignmentStart() >= region.getStart());
+                    } else {
+                        // mapped read.
+                        // the query gives us reads that overlap the region. We want only reads that *start* in those intervals.
+                        return (r.getStart() >= region.getStart());
+                    }
+                }
+
+
+            }));
+    }
+
+
+    // chr2:1-200 -> chr2:1-100,chr2:101-200
+    static public List<SimpleInterval> cutToShards(Iterable<SimpleInterval> intervals, int shardSize) {
```

Comment:
These sharding methods should either be private/protected, or fully documented and individually unit-tested.